Composable API: `replicate` and `DistributedState`

yhcharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89099
* __->__ #87649

This PR adds the first version of the `replicate()` composable API. For this prototype version, I try to reuse as much code from existing `DistributedDataParallel` as possible, and iterate on it in later changes. The basic idea of this prototype is:
- create a `ReplicateState` object. It internally uses a `ParameterList` module to hold all parameters of modules marked by `replicate()` API.
- create an internal `_ddp` object, which reuses existing `DistributedDataParallel` implementation, and wraps the `ParameterList` object
- install pre-forward and after-forward hooks on the root module, which calls methods of `_ddp` to run initialization and forward

yhcharles committed:
Draft of `replicate` and `DistributedState`

[ghstack-poisoned]

yhcharles committed:
Update on "Draft of `replicate` and `DistributedState`"




[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,66 @@
+from typing import List, Tuple
+
+
+class DistributedState:
```
nit: @mrshenli could we move 'DistributedState' to contract.py?

zhaojuanmao replied:
will this be cleaned up?

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,66 @@
+
+
+def replicate(
+    *modules: nn.Module, dist_state: ReplicateState = _default_state
```
nit: do we need to expose 'dist_state' to users?

yhcharles replied:
It's open for discussion. Maybe we don't need it and let all modules share the same `dist_state`.

yhcharles committed:
Update on "Draft of `replicate` and `DistributedState`"




[ghstack-poisoned]

yhcharles committed:
Update on "Draft of `replicate` and `DistributedState`"




[ghstack-poisoned]

yhcharles committed:
Update on "Draft of `replicate` and `DistributedState`"




[ghstack-poisoned]

yhcharles committed:
Update on "Draft of `replicate` and `DistributedState`"




[ghstack-poisoned]

yhcharles commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def replicate(
+    *modules: nn.Module, dist_state: ReplicateState = _default_state
```
This allows user to pass in a single module or multiple modules, for example:
`replicate(m)` or `replicate(m1, m2)`
In this case, `m1` and `m2` can be different modules in the tree.

yhcharles commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,65 @@
+class ReplicateState(DistributedState):
+    def __init__(self) -> None:
+        self.modules: List[nn.Module] = []
+        self.parameters: List[nn.Parameter] = []
```
The `replicate()` API can be called multiple times for different modules in a model. This `parameter` field is to keep a reference of all parameters of these modules. So that they can be managed together, for example bucketizing in DDP's all-reduce.

yhcharles commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,65 @@
+    def forward_pre_hook(
+        self, module: nn.Module, input: Tuple[torch.Tensor]
+    ) -> None:
+        if not self.has_initialized:
```
Do you mean hook on `self.modules` only, or recursively on all sub-modules of `self.modules`?

My current idea is former, `self.modules` only. Otherwise it may hurt performance. But it may change when we come to more details in implementation.

yhcharles commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,65 @@
+        self.parameters: List[nn.Parameter] = []
+        self.has_initialized: bool = False
+
+    def add_modules(self, *modules: nn.Module) -> None:
```
Sure, will do

yhcharles committed:
Update on "Draft of `replicate` and `DistributedState`"




[ghstack-poisoned]

yhcharles committed:
Update on "[WIP] Composable API: `replicate` and `DistributedState`"




[ghstack-poisoned]

yhcharles committed:
Update on "[WIP] Composable API: `replicate` and `DistributedState`"




[ghstack-poisoned]

yhcharles committed:
Update on "[WIP] Composable API: `replicate` and `DistributedState`"




[ghstack-poisoned]

yhcharles committed:
Update on "[WIP] Composable API: `replicate` and `DistributedState`"




[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,96 @@
+    dist_state.mark_modules(*modules)
+
+
+def mark_root_module(
```
feel a little confused about this API, I guess it will be private and called by replicate() later on?

yhcharles replied:
Please correct me if I'm wrong, I thought we have to ask users to call `mark_root()` to explicitly tell us the root module? Because `replicate()` itself has no way to find out the root.

zhaojuanmao replied:
I see, you want to install hooks on the root module only. that makes sense...

yhcharles commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,96 @@
+        ):  # managed by another API
+            return
+
+        self._param_list.extend(
+            param for param in module.parameters() if param.requires_grad
+        )
```
No, it's not. It's top-down search. The assumption is: whenever user marks a module, then this module and all its sub-modules (as long as them are not managed by other APIs) will be managed by `replicate()`. Does this sound good to you?

zhaojuanmao replied:
same here, if it is top-down search, the implementation seems to collect duplicate parameters?

yhcharles replied:
I'm not sure I get your point. Why do you think it collects duplicate parameters?
My idea is when it runs recursively on all sub-modules, it will first check if the sub-module is already marked by another API, if true, it will skip that sub-tree, so there won't be any duplicate?

zhaojuanmao replied:
module.parameters() will include children's parameters, for the case like this, b and d are marked to be replicated, e is marked to be sharded, what will 'self._param_list' be?

              root
            /     |     \
          a      b     c
                /    \
               d    e

zhaojuanmao replied:
I guess the best way is to write some unit tests to verify the parameters are collected as expected

yhcharles commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,96 @@
+            return
+
+        self.has_initialized = True
+        for module in self.modules:
```
User can call `replicate()` on multiple modules. They are stored in `self.modules`. And we only collect the parameters at the 1st forward during training.

zhaojuanmao replied:
one module could be child of another module in self.modules? seems the implementation here is collecting duplicated parameters?

yhcharles replied:
Yes, thanks for pointing it out. Updated in latest commit.

yhcharles commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,96 @@
+        for module in self.modules:
+            self._recursive_collect_params(module)
+
+        self._ddp = _ddp.DistributedDataParallel(self._param_list)
```
`self._param_list` is a `nn.ParameterList`, which is a sub-class of `nn.Module`.

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,96 @@
+
+
+def mark_root_module(
+    module: nn.Module, dist_state: ReplicateState = _default_state
```
mark_root_module() should be an API shared by all the composable APIs? so the dist_state type will not be constrained as ReplicateState?

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,96 @@
+        >>> module = nn.Linear(3, 3)
+        >>> replicate(module)
+    """
+    module.register_forward_pre_hook(dist_state.root_module_forward_pre_hook)
```
also rename 'register_forward_pre_hook' to 'register_replicate_forward_pre_hook'?

yhcharles replied:
`module.register_forward_pre_hook()` is the API provides by Autograd, we may not want to change that?

zhaojuanmao replied:
haha, right...

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,96 @@
+    """
+    module.register_forward_pre_hook(dist_state.root_module_forward_pre_hook)
+    # TODO(@yhcharles): fix type error
+    module.register_forward_hook(
```
same as above

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,96 @@
+
+    def _recursive_collect_params(self, module: nn.Module) -> None:
+        if (
+            getattr(module, "_distributed_state", None) is not None
```
also seems module._distributed_state is not None for modules that are marked as replicated at this point, so it always returns and did not fill in 'self._param_list'?

yhcharles replied:
Updated

mrshenli commented on torch/distributed/_composable/_ddp.py:
```diff
@@ -1166,6 +1167,15 @@ def forward(self, *inputs, **kwargs):
 
+    def forward(self, *inputs, **kwargs):
+        self.pre_forward(*inputs, **kwargs)
+        with torch.autograd.profiler.record_function(
```
this is not the same as existing DDP? As the `record_function` used to wrap pre_forward and post_forward here as well?

yhcharles replied:
That's true, it's not exactly the same.
Since it's now broken into 2 other functions, I added `record_function("DistributedDataParallel.pre_forward")` and `record_function("DistributedDataParallel.post_forward")` in each of them. Does this make sense?

mrshenli reviewed (commented):
please add test coverage

mrshenli commented on torch/distributed/_composable/_ddp.py:
```diff
@@ -1166,6 +1167,15 @@ def forward(self, *inputs, **kwargs):
             )
         return output
 
+    def forward(self, *inputs, **kwargs):
```
if this `_ddp.py` file is just for composable API, do we even need this forward method?

yhcharles replied:
Will clean it up soon

mrshenli reviewed (commented):
It will be great if we can have a clean up for the `_ddp.py` soon to remove unnecessary features. We usually don't leave duplicated code in PyTorch. I understand the intention is to move fast. But even in that case, let's try to only allow that for a short time window.

yhcharles committed:
Update on "[WIP] Composable API: `replicate` and `DistributedState`"




[ghstack-poisoned]

yhcharles committed:
Update on "[WIP] Composable API: `replicate` and `DistributedState`"


This PR adds the first version of the `replicate()` composable API. For this prototype version, I try to reuse as much code from existing `DistributedDataParallel` as possible, and iterate on it in later changes. The basic idea of this prototype is:
- create a `ReplicateState` object. It internally uses a `ParameterList` module to hold all parameters of modules marked by `replicate()` API.
- create an internal `_ddp` object, which reuses existing `DistributedDataParallel` implementation, and wraps the `ParameterList` object
- install pre-forward and after-forward hooks on the root module, which calls methods of `_ddp` to run initialization and forward


[ghstack-poisoned]

yhcharles committed:
Update on "[WIP] Composable API: `replicate` and `DistributedState`"


This PR adds the first version of the `replicate()` composable API. For this prototype version, I try to reuse as much code from existing `DistributedDataParallel` as possible, and iterate on it in later changes. The basic idea of this prototype is:
- create a `ReplicateState` object. It internally uses a `ParameterList` module to hold all parameters of modules marked by `replicate()` API.
- create an internal `_ddp` object, which reuses existing `DistributedDataParallel` implementation, and wraps the `ParameterList` object
- install pre-forward and after-forward hooks on the root module, which calls methods of `_ddp` to run initialization and forward


[ghstack-poisoned]

yhcharles committed:
Update on "[WIP] Composable API: `replicate` and `DistributedState`"


This PR adds the first version of the `replicate()` composable API. For this prototype version, I try to reuse as much code from existing `DistributedDataParallel` as possible, and iterate on it in later changes. The basic idea of this prototype is:
- create a `ReplicateState` object. It internally uses a `ParameterList` module to hold all parameters of modules marked by `replicate()` API.
- create an internal `_ddp` object, which reuses existing `DistributedDataParallel` implementation, and wraps the `ParameterList` object
- install pre-forward and after-forward hooks on the root module, which calls methods of `_ddp` to run initialization and forward


[ghstack-poisoned]

yhcharles committed:
Update on "[WIP] Composable API: `replicate` and `DistributedState`"


This PR adds the first version of the `replicate()` composable API. For this prototype version, I try to reuse as much code from existing `DistributedDataParallel` as possible, and iterate on it in later changes. The basic idea of this prototype is:
- create a `ReplicateState` object. It internally uses a `ParameterList` module to hold all parameters of modules marked by `replicate()` API.
- create an internal `_ddp` object, which reuses existing `DistributedDataParallel` implementation, and wraps the `ParameterList` object
- install pre-forward and after-forward hooks on the root module, which calls methods of `_ddp` to run initialization and forward


[ghstack-poisoned]

zhaojuanmao commented on test/distributed/_composable/test_replicate.py:
```diff
@@ -0,0 +1,100 @@
+
+    def test_replicate(self):
+        dist.init_process_group(
+            backend="gloo",
```
let's test on both gloo and nccl

zhaojuanmao commented on test/distributed/_composable/test_replicate.py:
```diff
@@ -0,0 +1,100 @@
+        local_batch_size = 1
+        global_batch_size = self.world_size * local_batch_size
+        model, input, target = self._prepare_module(global_batch_size)
+        replicate_model = mark_root_module(replicate(deepcopy(model)))
```
let's add more test cases: 
1. replicate one submodule instead of the root module
2. replicate  more than one submodules

zhaojuanmao replied:
also a test case where there are some submodules of the replicated local root module are annotated by fully_shard()

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,107 @@
+        for module in modules:
+            self.modules.append(module)
+            replicate.state(module)._distributed_state = self
+            replicate.state(module)._params_collected = False
```
wondering how other states in ddp constructor are populated?

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,107 @@
+        for module in self.modules:
+            self._recursive_collect_params(module)
+
+        self._ddp = _ddp.DistributedDataParallel(self._param_list)
```
oh, all the states in ddp constructor is owned in self._ddp...

assume we will not use this soon, as it is still monkey patching. instead, we will make all the states owned by replicate.state() object?

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,107 @@
+        self, module: nn.Module, input: Tuple[torch.Tensor]
+    ) -> None:
+        self.init_helper()
+        self._ddp.pre_forward()
```
same as above, we need to make pre_forward() accept a replicate.state() object once get rid of self._ddp?

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,107 @@
+@contract
+def replicate(
+    module: nn.Module,  # NOTE: contract now supports single module only
+    dist_state: ReplicateState = _default_state,
```
dist_state is used internally, could we remove it from user facing API here? 

also, replicate needs to have similar argument as exiting DDP API, can work with important features that DDP provided, like static_graph, gradient_as_bucket_view, etc

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,107 @@
+
+
+def mark_root_module(
+    module: nn.Module, dist_state: ReplicateState = _default_state
```
same as above, not exposing dist_state to users

zhaojuanmao commented on torch/distributed/_composable/replicate.py:
```diff
@@ -0,0 +1,107 @@
+        modules (torch.nn.Module): root module
+
+    Example::
+        >>> module = nn.Linear(3, 3)
+        >>> replicate(module)
```
nit: the example does not reflect the 'mark_root_module'

yhcharles committed:
Update on "[WIP] Composable API: `replicate` and `DistributedState`"


This PR adds the first version of the `replicate()` composable API. For this prototype version, I try to reuse as much code from existing `DistributedDataParallel` as possible, and iterate on it in later changes. The basic idea of this prototype is:
- create a `ReplicateState` object. It internally uses a `ParameterList` module to hold all parameters of modules marked by `replicate()` API.
- create an internal `_ddp` object, which reuses existing `DistributedDataParallel` implementation, and wraps the `ParameterList` object
- install pre-forward and after-forward hooks on the root module, which calls methods of `_ddp` to run initialization and forward


[ghstack-poisoned]

zhaojuanmao reviewed (approved):
stamp to get this version backed up. 

Synced up offline, following PRs will be sent out: 
1. remove mark_root_module()
2. clean up replicate() constructor without distState argument and add important arguments such as static_graph, find_unused_parameter, etc
3. clean up self._ddp
4. add more tests like interact with fully_shard.py

yhcharles commented:
@pytorchbot merge
